File: packages/crypto/src/keys.ts
import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as nodeSign,
  verify as nodeVerify,
} from 'node:crypto';
import { digestBytes } from '@act/core';

export interface KeyPair {
  /** ACT key identifier: "ed25519:<sha-256 fingerprint of the raw public key, hex>". */
  keyId: string;
  /** Raw 32-byte Ed25519 public key, base64-encoded. */
  publicKey: string;
  /** Raw 32-byte Ed25519 private key seed, base64-encoded. Never log or persist in plaintext outside a key store. */
  privateKey: string;
}

function base64UrlToBase64(b64url: string): string {
  const b64 = b64url.replace(/-/g, '+').replace(/_/g, '/');
  const padded = b64 + '='.repeat((4 - (b64.length % 4)) % 4);
  return padded;
}

function base64ToBase64Url(b64: string): string {
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/** Generates a fresh Ed25519 key pair. */
export function generateKeyPair(): KeyPair {
  const { publicKey, privateKey } = generateKeyPairSync('ed25519');
  const publicJwk = publicKey.export({ format: 'jwk' }) as { x: string };
  const privateJwk = privateKey.export({ format: 'jwk' }) as { d: string };
  const publicKeyB64 = base64UrlToBase64(publicJwk.x);
  const privateKeyB64 = base64UrlToBase64(privateJwk.d);
  return {
    keyId: keyIdFor(publicKeyB64),
    publicKey: publicKeyB64,
    privateKey: privateKeyB64,
  };
}

/** Derives the ACT key_id ("ed25519:<hex fingerprint>") for a raw base64-encoded public key. */
export function keyIdFor(publicKeyBase64: string): string {
  const digest = digestBytes(Buffer.from(publicKeyBase64, 'base64'));
  return `ed25519:${digest.replace(/^sha-256:/, '')}`;
}

function publicKeyObjectFrom(publicKeyBase64: string) {
  const x = base64ToBase64Url(publicKeyBase64);
  return createPublicKey({ key: { kty: 'OKP', crv: 'Ed25519', x }, format: 'jwk' });
}

function privateKeyObjectFrom(privateKeyBase64: string, publicKeyBase64: string) {
  const x = base64ToBase64Url(publicKeyBase64);
  const d = base64ToBase64Url(privateKeyBase64);
  return createPrivateKey({ key: { kty: 'OKP', crv: 'Ed25519', x, d }, format: 'jwk' });
}

/** Signs raw bytes with an Ed25519 private key, returning a base64-encoded signature. */
export function signBytes(
  privateKeyBase64: string,
  publicKeyBase64: string,
  message: Uint8Array,
): string {
  const key = privateKeyObjectFrom(privateKeyBase64, publicKeyBase64);
  const signature = nodeSign(null, message, key);
  return signature.toString('base64');
}

/** Verifies a base64-encoded Ed25519 signature over raw bytes. */
export function verifyBytes(
  publicKeyBase64: string,
  message: Uint8Array,
  signatureBase64: string,
): boolean {
  try {
    const key = publicKeyObjectFrom(publicKeyBase64);
    return nodeVerify(null, message, key, Buffer.from(signatureBase64, 'base64'));
  } catch {
    return false;
  }
}

File: sdks/python/src/act_sdk/crypto/keys.py
"""Ports `packages/crypto/src/keys.ts`.

Uses the `cryptography` package's Ed25519 primitives (there is no Ed25519
support in the Python standard library) -- signatures are deterministic
per RFC 8032, so a fixed keypair and message byte-for-byte match the
signature any other conformant implementation (including
`packages/crypto`) produces for the same inputs; see
`conformance/vectors/signatures.json`.
"""

from __future__ import annotations

import base64
from dataclasses import dataclass

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    NoEncryption,
    PrivateFormat,
    PublicFormat,
)

from ..core.digest import digest_bytes


@dataclass(frozen=True)
class KeyPair:
    """ACT key identifier + raw Ed25519 key material, base64-encoded."""

    key_id: str
    public_key: str
    private_key: str


def generate_key_pair() -> KeyPair:
    """Generates a fresh Ed25519 key pair."""
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()
    private_key_b64 = base64.b64encode(
        private_key.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())
    ).decode("ascii")
    public_key_b64 = base64.b64encode(
        public_key.public_bytes(Encoding.Raw, PublicFormat.Raw)
    ).decode("ascii")
    return KeyPair(
        key_id=key_id_for(public_key_b64),
        public_key=public_key_b64,
        private_key=private_key_b64,
    )


def key_id_for(public_key_base64: str) -> str:
    """Derives the ACT key_id ("ed25519:<hex fingerprint>") for a raw base64-encoded public key."""
    digest = digest_bytes(base64.b64decode(public_key_base64))
    return f"ed25519:{digest.split(':', 1)[1]}"


def sign_bytes(private_key_base64: str, message: bytes) -> str:
    """Signs raw bytes with an Ed25519 private key, returning a base64-encoded signature.

    Unlike the TypeScript SDK's `signBytes`, this does not also take the
    public key -- Ed25519 private key material alone determines the
    signature; the public-key parameter over there is an artifact of how
    Node's WebCrypto/JWK key import works, not a cryptographic requirement.
    """
    private_key = Ed25519PrivateKey.from_private_bytes(base64.b64decode(private_key_base64))
    signature = private_key.sign(message)
    return base64.b64encode(signature).decode("ascii")


def verify_bytes(public_key_base64: str, message: bytes, signature_base64: str) -> bool:
    """Verifies a base64-encoded Ed25519 signature over raw bytes."""
    try:
        public_key = Ed25519PublicKey.from_public_bytes(base64.b64decode(public_key_base64))
        public_key.verify(base64.b64decode(signature_base64), message)
        return True
    except (InvalidSignature, ValueError):
        return False
